second_pass.c

```diff
@@ -142,258 +142,237 @@ void read_line2(char *line, FILE *ob_file, FILE *ent_file, FILE *ext_file, int *
 		int instindex = is_inst(line);
-		int addressing1 = 0, addressing2 = 0;
-		int i, data = 0, reg_src = 0, reg_dst = 0;
+		int addressing1 = 0, addressing2 = 0, addressing3 = 0;
+		int i, data = 0, reg_src = 0, reg_dst = 0,immed=0;
 		char label[MAX_LABEL_LEN];
 		line = next_word(line);
-		int insttype=check_inst_type(instindex);
+		char instname = check_inst_name(check_inst_type(instindex));
 
-		/* checks if it is R_ARITHMETHIC instruction */
-		if(insttype == R_ARITHMETHIC){
-			to_inst()
-		}
-		
-		/* if the instruction needs two operands */
-		else if(instindex < NOT_INST || instindex == LEA_INST)
-		{
+		switch(instname){
+			case 'add':
 			addressing1 = check_addressing(line, error);
 			addressing2 = check_addressing(to_comma(line), error);
-			to_inst(INST_TYPE, instindex
-, addressing1, addressing2, data, ABSOLUTE, ic);
-			(*ic)++;
+			addressing3 = check_addressing(to_comma(to_comma(line,error),error));
+			to_inst(0,addressing1,addressing2,addressing3,1,IRELEVANT,IRELEVANT,ic);
+			(*ic++);
+			break;
 
-			/* if first operand is immidiate number */
-			if(addressing1 == IMM_ADDRESS)
-			{
-				symbol_node * node;
-				char key[MAX_MACRO_LEN];
-				int i = 0;
-				line++;
-			
-			/* if first operand is a label */
-			if(addressing1 == LABEL_ADDRESS)
-			{
-				for(i=0;*line!=',';i++)
-				{
-					label[i] = *line;
-					line++;
-				}
-				
-				label[i] = '\0';
-				symbol = search_sym(label);
-				
-				if(symbol == NULL)
-				{
-					*error = NDEF_LABEL_ERROR;
-					return;
-				}
-				
-				/* if the label is external */
-				else if(symbol->ext_flag == TRUE)
-				{
-					to_ent_ext(label, *ic, symbol->ext_flag);
-					to_inst(LABEL_TYPE, instindex
-		, addressing1, addressing2, FALSE, EXTERNAL, ic);
-					(*ic)++;
-				}
-				
-				else
-				{
-					to_inst(LABEL_TYPE, instindex
-		, addressing1, addressing2, symbol->address, RELOCATABLE, ic);
-					(*ic)++;
-				}
-			}
-			
-			/* if first operand is an index */
-			if(addressing1 == INDEX_ADDRESS)
-			{
-				/*check if macro */
-				char key[MAX_MACRO_LEN];
-				int j=1;
-				symbol_node * node;
-
-				for(i=0;*line!='[';i++)
-				{
-					label[i] = *line;
-					line++;
-				}
-				label[i] = '\0';
-
-				symbol = search_sym(label);
-				while(line[j] != ']' && line[j] != '\0')
-				{
-					key[j-1] = line[j];
-					j++;
-				}
-				key[j-1] = '\0';
-
-				if((node = search_sym(key)) != NULL)
-				{
-					*(line+1) = node->address + '0';
-				}
-
-				if(symbol == NULL)
-				{
-					*error = NDEF_LABEL_ERROR;
-					return;
-				}
-
-			/* if the index is external */
-			else if(symbol->ext_flag == TRUE)
-			{
-				to_ent_ext(label, *ic, symbol->ext_flag);
-				to_inst(LABEL_TYPE, instindex
-	, addressing1, addressing2, FALSE, EXTERNAL, ic);
-				(*ic)++;
-			}
-			else
-			{
-				to_inst(LABEL_TYPE, instindex
-	, addressing1, addressing2, symbol->address, RELOCATABLE, ic);
-				(*ic)++;
-			}
-			line++;
+			case 'sub':
+			addressing1 = check_addressing(line, error);
```

Comment:
those 3 lines are doubled per each one of those 30 cases .

create the parameters you would like to add to the instruction node in the above:
int rd,rs,rt,opcode,funct;
on each case fulfill those parameters, for example :

case "sub":
{
opcode = 3;
rd = 2;
rt = 5;
immed = 1;
}

when you finish the switch case, call to to_inst(opcode,rt,rs,immed) etc..
this will also remove one line from each case.

Reply:
but if im doing this, i will need to define more variables in each case, it will just add more lines..